package/scripts/postinstall

```diff
@@ -43,20 +43,42 @@ then
 fi
 
 # create missing directories
-sudo mkdir -vp Cellar Frameworks etc include lib opt sbin share var/homebrew/linked
+mkdir -vp Cellar Frameworks etc include lib opt sbin share var/homebrew/linked
 
-# set permissions
+# get valid logged in user
 logged_in_user=$(echo "show State:/Users/ConsoleUser" | scutil | awk '/Name :/ { print $3 }')
-group=$(id -gn "${logged_in_user}")
+
+# optionally pass an environment variable to `installer` to define user
+if [[ -n "${HOMEBREW_PKG_USER:-}" ]]
+then
+  logged_in_user="${HOMEBREW_PKG_USER}"
+elif [[ "${logged_in_user}" =~ _mbsetupuser|loginwindow|root ]] || [[ -z "${logged_in_user}" ]]
+then
+  # get last created account UID over 500
+  logged_in_user=$(dscl . list /Users UniqueID | awk '$2 > 500 { print $1 }' | tail -n 1)
+  # no valid accounts exist yet, use root temporarily
+  if [[ -z "${logged_in_user}" ]]
+  then
+    logged_in_user="root"
+  fi
```

Comment:
Agreed assigning `root` as owner is not the solution here. How would we like the install to fail? At this point in the install files are already on disk in `/opt/homebrew`. If using this conditional to bail halfway through users might assume the package "failed successfully" or is in a healthy enough state to proceed. Or Homebrew could be broken enough they don't know what to do. 

I think the best solution is to fail early. Since we're in agreement installs with no user created yet shouldn't be supported, let's move or add `logged_in_user` logic to a preinstall script. Then a precondition check can fail before the package lays down files. Thoughts?

```bash
  # no valid accounts exist yet, use root temporarily
  if [[ -z "${logged_in_user}" ]]
  then
    logged_in_user="root"
  fi
```

The above would change to something like...

```bash
  # no valid accounts exist yet, use root temporarily
  if [[ -z "${logged_in_user}" ]]
  then
    echo "No valid user to install Homebrew. Exiting."
    exit 1
  fi
```

https://github.com/Homebrew/brew/blob/709b4680e76ea9fddcf4eee0b869dc959de07245/package/scripts/postinstall#L51-L64

Reply:
> Since we're in agreement installs with no user created yet shouldn't be supported, let's move or add `logged_in_user` logic to a preinstall script.

Agreed 👍🏻 

I think a `Distribution.xml` `<script>` check, if possible, may be even better as it allows for a nicer error message flow.

Reply:
Tried to shell out in `Distribution.xml` using `run` (https://developer.apple.com/documentation/installer_js/system/1812364-run), but Apple has deprecated it in `productbuild`. Instead I opted for a preinstall script. Though the user experience isn't quite as nice, this error should never be encountered by a logged in user installing the package manually. The other option is to use `users.desktopSessionCount` (https://developer.apple.com/documentation/installer_js/system/1812277-users_desktopsessioncount) to determine if a user is logged in, but still holding out hope we can reintroduce login window support.

```
productbuild: [WARNING]: Specifying [allow-external-scripts='true'] is deprecated.
           Having this in your distribution will break the installability
           of this product in an upcoming release of macOS.
           Apple reserves the right to reject notarization of any product with this option.
```
<img width="620" alt="Screenshot 2023-10-04 at 12 37 46 PM" src="https://github.com/Homebrew/brew/assets/17690356/2edc76ca-ad2d-4fb7-9e4e-50eba06c699f">